src/Chainweb/Pact/PactService.hs

```diff
@@ -573,90 +598,143 @@ validateHashes bHeader pData miner transactions =
-    cenv <- restore $ restoreCheckpointer target caller
-    try (restore (act cenv)) >>= \case
-        Left e -> discardTx >> throwM @_ @SomeException e
-        Right (Discard !result) -> discardTx >> return result
-        Right (Save header !result) -> saveTx header >> return result
+withCheckpointerWithoutRewind target caller act = do
+    checkPointer <- getCheckpointer
+    logInfo $ "restoring (with caller " <> caller <> ") " <> sshow target
+
+    -- check requirement that this must be called within a batch
+    unlessM (asks _psIsBatch) $
+        error $ "Code invariant violation: withCheckpointer called by " <> caller <> " outside of batch. Please report this as a bug."
+    -- we allow exactly one nested call of 'withCheckpointer', which is used
+    -- during fastforward in 'rewindTo'.
+    unlessM ((<= 1) <$> asks _psCheckpointerDepth) $ do
+        error $ "Code invariant violation: to many nested calls of withCheckpointer. Please report this as a bug."
+
+    local (over psCheckpointerDepth succ) $ mask $ \restore -> do
+        cenv <- restore $ do
+            r <- liftIO $ _cpRestore checkPointer target
+            return r
+
+        try (restore (act cenv)) >>= \case
+            Left e -> discardTx >> throwM @_ @SomeException e
+            Right (Discard !result) -> discardTx >> return result
+            Right (Save header !result) -> saveTx header >> return result
   where
     discardTx = finalizeCheckpointer _cpDiscard
     saveTx !header = do
+        -- TODO: _cpSave is a complex call. If any thing in there throws
+        -- an exception it would result in a pending tx.
         finalizeCheckpointer (flip _cpSave $ _blockHash header)
-        modify' $ set psStateValidated $ Just header
+        modify' $ set psStateValidated (Just header)
+        setParentHeader (caller <> ".withCheckpointerWithoutRewind") (ParentHeader header)
 
 -- | 'withCheckpointer' but using the cached parent header for target.
+--
 withCurrentCheckpointer
     :: PayloadCasLookup cas
     => String
     -> (PactDbEnv' -> PactServiceM cas (WithCheckpointerResult a))
     -> PactServiceM cas a
 withCurrentCheckpointer caller act = do
-    ph <- _parentHeader <$> use psParentHeader
+    ParentHeader ph <- syncParentHeader "withCurrentCheckpointer"
     let target = Just (succ $ _blockHeight ph, _blockHash ph)
-    withCheckpointer target caller act
+    withCheckpointer (Just 0) target caller act
 
--- | Same as 'withCheckpointer' but rewinds the checkpointer state to the
+-- | Execute an action in the context of an @Block@ that is provided by the
+-- checkpointer. The checkpointer is rewinded and restored to the state to the
 -- provided target.
 --
-withCheckpointerRewind
+-- The result of the inner action indicates whether the resulting checkpointer
+-- state should be discarded or saved.
+--
+-- If the inner action throws an exception the checkpointer state is discarded.
+--
+withCheckpointer
```

Comment:
Might want to leave this `withCheckpointerRewind` since at this time calling code needs to be aware of semantics (unfortunately).

Reply:
Yes, renaming it to `withCheckpointerRewind` makes sense.